README.md
# Dev.Opera

We’ve rebuilt [Dev.Opera](https://dev.opera.com/) as a static site, powered by Jekyll, Sass, and Grunt. In this repository, you find all the source files and content to build the site, [make improvements](CONTRIBUTING.md#code-improvements), and [submit new articles](CONTRIBUTING.md#article-suggestions--contributions).

## Environment setup

1. Install Bundler by running `gem install bundler` on the command line.
2. Download and install [Node.js](http://nodejs.org/) or [io.js](https://iojs.org/) with the default installer options.
3. Clone the project by running `git clone git@github.com:operasoftware/devopera.git` and `cd` into the `devopera` folder.
4. Install Jekyll, Sass, and all needed gems by running `bundle install`.
5. Run `npm install` inside the `devopera` folder.
6. Run `npm install bower -g` to install Bower globally.
7. Run `bower install`.
8. Run `npm install grunt-cli -g` to install Grunt globally.

Please install [EditorConfig](http://editorconfig.org/#download) plugin for your editor to keep code style declared in the [`.editorconfig`](.editorconfig) file.

## Environment update

If you’ve set up all the above at some point in the past, and want to update your installation, these are the steps to follow:

1. Run `sudo gem update` to install the latest Ruby gems.
2. Run `npm install` & `bower install` inside the `devopera` folder.

## Development

To start developing Dev.Opera run `grunt`, it would take up to 1 minute to:

1. Build site in a limited mode with only latest 150 posts.
2. Start local server on `http://0.0.0.0:33310` address.
3. Open it in your default browser.
4. Launch a `watch` task for all files.

So now every time you change project files, Dev.Opera will be rebuilt and you browser will update. For HTML and Markdown files it will still take up to 1 minute, for CSS it will be much faster (please request the same fast track for JS or anything else if needed).

# Full build and deploy

To **build** a complete Dev.Opera, run `grunt build`: it will take a while to build the full site with all posts included in the `_site` folder. To **deploy** (sync build with remote server), run `grunt deploy`.

## Notes

- All instructions are for Unix (OS X, Linux, etc.) because Jekyll is not compatible with Windows.
- You may require write access for all commands mentioned above. Type `sudo`, space and command if needed.


_extensions/history.md
---
title: Working With the Browser History
authors:
- shwetank-dixit
source: http://developer.chrome.com/extensions/history.html
license: cc-by-3.0
---

## Introduction

The History API allows an extension to access and edit the user’s browsing history. This can come in handy in some situations, for example analyzing which are the most visited pages by the user, maintaining a list of closed tabs to re-open later, etc.

The first thing to note is to declare it in the `permissions` field in the mainifest file, like so:

	"permissions": [
		"history"
	]

## Transition Types

_Transition Type_ is a term used to describe the way in which a particular URL has been navigated to. For example, a transition type of `link` will refer to the fact that the URL was navigated through by clicking on a hyperlink in a web page. The transition type of `typed` would mean that the user went to that page by explicitly typing that URL in the address bar, for instance. The full list of transition types are below:

<figure block="figure">
<table>
<tr>
	<th>Transition type</th>
	<th>Description </th>
</tr>
<tr id="tt_link">
	<td><code>link</code></td>
	<td>The user got to this page by clicking a link on another page.</td>
</tr>
<tr id="tt_typed">
	<td><code>typed</code></td>
	<td>The user got this page by typing the URL in the address bar. Also used for other explicit navigation actions. See also <a href="#tt_generated"><code>generated</code></a>, which is used for cases where the user selected a choice that didn’t look at all like a URL.</td>
</tr>
<tr id="tt_auto_bookmark">
	<td><code>auto_bookmark</code></td>
	<td>The user got to this page through a suggestion in the UI — for example, through a menu item.</td>
</tr>
<tr id="tt_auto_subframe">
	<td><code>auto_subframe</code></td>
	<td>Subframe navigation. This is any content that is automatically loaded in a non-top-level frame. For example, if a page consists of several frames containing ads, those ad URLs have this transition type. The user may not even realize the content in these pages is a separate frame, and so may not care about the URL (see also <a href="#tt_manual_subframe"><code>manual_subframe</code></a>).</td>
</tr>
<tr id="tt_manual_subframe">
	<td><code>manual_subframe</code></td>
	<td>For subframe navigations that are explicitly requested by the user and generate new navigation entries in the back/forward list. An explicitly requested frame is probably more important than an automatically loaded frame because the user probably cares about the fact that the requested frame was loaded.</td>
</tr>
<tr id="tt_generated">
	<td><code>generated</code></td>
	<td>The user got to this page by typing in the address bar and selecting an entry that did not look like a URL. For example, a match might have the URL of a Google search result page, but it might appear to the user as “Search Google for…”. These are not quite the same as <a href="#tt_typed"><code>typed</code></a> navigations because the user didn’t type or see the destination URL. See also <a href="#tt_keyword"><code>keyword</code></a>.</td>
</tr>
<tr id="tt_auto_toplevel">
	<td><code>auto_toplevel</code></td>
	<td>The page was specified in the command line or is the start page.</td>
</tr>
<tr id="tt_form_submit">
	<td><code>form_submit</code></td>
	<td>The user filled out values in a form and submitted it. Note that in some situations — such as when a form uses script to submit contents — submitting a form does not result in this transition type.</td>
</tr>
<tr id="tt_reload">
	<td><code>reload</code></td>
	<td>The user reloaded the page, either by clicking the reload button or by pressing Enter in the address bar. Session restore and Reopen closed tab use this transition type, too.</td>
</tr>
<tr id="tt_keyword">
	<td><code>keyword</code></td>
	<td>The URL was generated from a replaceable keyword other than the default search provider.
		See also <a href="#tt_keyword_generated"><code>keyword_generated</code></a>.</td>
</tr>
<tr id="tt_keyword_generated">
	<td><code>keyword_generated</code></td>
	<td>Corresponds to a visit generated for a keyword. See also <a href="#tt_keyword"><code>keyword</code></a>.</td>
</tr>
</table>
</figure>

Each and every visit to any URL will be recorded in the history as a [`VisitItem`](https://developer.chrome.com/extensions/history#type-VisitItem). This item, besides holding detailed information about every visit, such as the URL visited, the time at which the visit occurred etc, will also hold information regarding the _transition type_, that is, how the user reached the webpage. Thus each and every visit to a web page will have a _transition type_ associated with it.

## Accessing the history behind the currently active tab

Let’s say you wanted to know how many times the URL in the currently active tab has been accessed. We can take a look at the methods in the History API and build something to find it out.

In our example, we’ll create a _browser action_, which when clicked, will update its badge with the number of times it has been accessed by the user. Let’s see the code for it, which is in the background script.

	var count = 0;

	chrome.browserAction.onClicked.addListener(function (tab) {
		chrome.history.getVisits({
			'url': tab.url
		}, function (visitItem) {
			count = visitItem.length;
			chrome.browserAction.setBadgeBackgroundColor({
				color: '#ff0000'
			});
			chrome.browserAction.setBadgeText({
				text: '' + count
			});
			console.log('The user has visited ' +
				url + ' ' +
				count + ' times.');
		});
	});


Gruntfile.js
module.exports = function(grunt) {

	require('load-grunt-tasks')(grunt);
	require('time-grunt')(grunt);

	grunt.initConfig({
		sass: {
			task: {
				files: [{
					expand: true,
					cwd: 'styles',
					src: [
						'screen.scss',
						'search.scss'
					],
					dest: 'styles',
					ext: '.css'
				}]
			}
		},
		cssmin: {
			task: {
				files: [{
					expand: true,
					cwd: 'styles',
					src: '*.css',
					dest: 'styles'
				}]
			}
		},
		autoprefixer: {
			task: {
				src: 'styles/*.css'
			}
		},
		clean: [
			'_site'
		],
		jekyll: {
			full: {
				//
			},
			limit: {
				options: {
					limit_posts: 150
				}
			}
		},
		htmlmin: {
			html: {
				options: {
					removeComments: true,
					collapseWhitespace: true
				},
				expand: true,
				cwd: '_site/',
				src: [
					'**/index.html',
					'errors/*.html'
				],
				dest: '_site/'
			},
			xml: {
				options: {
					collapseWhitespace: true,
					keepClosingSlash: true
				},
				expand: true,
				cwd: '_site/feed/',
				src: '**/index.xml',
				dest: '_site/feed/'
			}
		},
		beml: {
			options: {
				elemPrefix: '__',
				modPrefix: '--'
			},
			files: {
				expand: true,
				cwd: '_site/',
				src: [
					'**/index.html',
					'errors/*.html'
				],
				dest: '_site/'
			}
		},
		connect: {
			task: {
				options: {
					base: '_site',
					port: 33310,
					livereload: true,
					open: true
				}
			}
		},
		watch: {
			styles: {
				files: 'styles/*.scss',
				tasks: [
					'styles'
				]
			},
			limit: {
				files: [
					'**',

					'!.git/**',
					'!.sass-cache/**',

					'!_scripts/**',
					'!_site/**',

					'!CONTRIBUTING.md',
					'!LICENSE.md',
					'!README.md',

					'!Gruntfile.js',
					'!node_modules/**',
					'!package.json',
					'!install.sh',
					'!styles/*',

					'!{articles,blog,tv,extensions}/**',
						'{articles,blog,tv,extensions}/index.html',
						'{articles,blog,tv,extensions}/**/_posts/*.md'
				],
				tasks: [
					'jekyll:limit',
					'html',
					'styles'
				]
			},
			livereload: {
				options: {
					livereload: true
				},
				files: [
					'_site/**/*.html',
					'_site/styles/*.css'
				]
			}
		},
		copy: {
			task: {
				files: [{
					expand: true,
					src: 'styles/*.css',
					dest: '_site/'
				}]
			}
		},
		replace: {
			htaccess: {
				src: '.htaccess',
				dest: '_site/',
				replacements: [{
					// Redirect all traffic to https://dev.opera.com/*
					from: 'RewriteEngine On\n\n',
					to: 'RewriteEngine On\n\nRewriteCond %{HTTP_HOST} !^dev\\.opera\\.com [NC]\nRewriteRule ^(.*)$ https://dev.opera.com%{REQUEST_URI} [R=301,L]\n\nRewriteCond %{HTTPS} off\nRewriteRule ^(.*)$ https://dev.opera.com%{REQUEST_URI} [R=301,L]\n\n'
				}]
			},
			htmlLinks: {
				src: [
					'_site/**/*.html'
				],
				overwrite: true,
				replacements: [{
					from: /(<link rel="stylesheet" href="\/styles\/)(screen)(\.css">)/g,
					to: '$1' + require('hash-files').sync({
						files: [
							'_site/styles/screen.css'
						]
					}).substring(0, 8) + '$3'
				}]
			},
			cachedFiles: {
				src: [
					'_site/service-worker.js'
				],
				overwrite: true,
				replacements: [{
					from: /(const HASH = ')(';)/g,
					to: '$1' + require('hash-files').sync({
						files: [
							'_site/styles/screen.css',
							'_site/images/github.svg',
							'_site/images/logo.png',
							'_site/images/logo@2x.png',
							'_site/scripts/highlight.js',
							'_site/scripts/salvattore.js'
						]
					}).substring(0, 8) + '$2'
				},{
					from: /('\/styles\/)(screen)(\.css',)/g,
					to: '$1' + require('hash-files').sync({
						files: [
							'_site/styles/screen.css'
						]
					}).substring(0, 8) + '$3'
				}]
			}
		},
		rename: {
			cachedStyles: {
				src: '_site/styles/screen.css',
				dest: '_site/styles/' + require('hash-files').sync({
					files: [
						'_site/styles/screen.css'
					]
				}).substring(0, 8) + '.css'
			}
		},
		rsync: {
			options: {
				args: [
					'--checksum', // skip based on checksum, not mod-time & size
					'--chmod=ug=rwX,o=rX', // `chmod` new files
					'--compress', // compress data during the transfer
					// Note: the next line is commented out to avoid deleting the
					// `extension-docs` directory, which is generated from a separate
					// repository at the moment. As soon as the extension docs are merged
					// with the main `devopera` repository, we should re-enable it.
					//'--delete', // delete extraneous files on the receiving side
					'--human-readable', // output numbers in a human-readable format
					'--itemize-changes', // output a change-summary for all updates
					'--omit-dir-times', // …except for directories
					'--progress', // show progress during transfer
					'--stats', // show some file-transfer stats
					'--times', // preserve modification times…
					'--verbose' // increase verbosity
				],
				exclude: ['.DS_Store'],
				recursive: true
			},
			deploy: {
				options: {
					src: '_site/',
					dest: '/var/www/html/',
					host: '54.213.240.91'
				}
			}
		}
	});

	grunt.registerTask('styles', [
		'sass',
		'autoprefixer',
		'cssmin',
		'copy'
	]);

	grunt.registerTask('html', [
		'htmlmin',
		'beml'
	]);

	grunt.registerTask('cache', [
		'replace:htmlLinks',
		'replace:cachedFiles',
		'rename:cachedStyles'
	]);

	grunt.registerTask('default', [
		'clean',
		'jekyll:limit',
		'html',
		'styles',
		'cache',
		'connect',
		'watch'
	]);

	grunt.registerTask('ci', [
		'clean',
		'styles',
		'jekyll:full',
		'html',
		'cache'
	]);

	grunt.registerTask('build', [
		'ci',
		'connect:task:keepalive'
	]);

	grunt.registerTask('deploy', [
		'replace:htaccess',
		'rsync'
	]);

};
